### ISSUE TITLE
AWSXRayRecorder.Instance.GetEntity() doesn't return most recently opened subsegment

### ISSUE THREAD

**Issue by williamdenton** [aws/aws-xray-sdk-dotnet#84]
I am writing an xray wrapper for Postgres so I can capture the SQL being executed on RDS.

I'm using the Miniprofiler core classes (but not mini-profiler it self) as it provides the necessary hook points and wrapping classes to get this going.

I have implemented by own IDbProfiler based on this interface
https://github.com/MiniProfiler/dotnet/blob/master/src/MiniProfiler.Shared/Data/IDbProfiler.cs

in `ExecuteStart()` i start a segment and add the sql info. 
```cs
var recorder = AWSXRayRecorder.Instance;
recorder.BeginSubsegment("SQL");
recorder.AddSqlInformation("sanitized_query", dbCommand.CommandText);
...
```

in `ExecuteFinish()` i close the subsegment
```cs
AWSXRayRecorder.Instance.EndSubsegment();
```

Between Start and Finish Entity framework is executing an async db query, but no other profiling takes place.

In ExecuteStart() if I add `AWSXRayRecorder.Instance.GetEntity()` i get the newly added subsegment back as expected.
In `ExecuteFinish()` the segment returned by `AWSXRayRecorder.Instance.GetEntity()` is the root `Segment` not the `Subsegment` I opened in start execute. Inspecting the the root segment shows that there is a single subsegment in the `Subsegments` collection with the name "SQL", and `IsInProgress` is true. If I extract that segment and set it as the current entity I can end the subsegment and my profiling works as expected.
I'm using code like this (which i think is a gross work around):
```cs
var recorder = AWSXRayRecorder.Instance;
var segment = recorder.GetEntity();
if (segment is Amazon.XRay.Recorder.Core.Internal.Entities.Segment rootSegment) {
    var sqlSegment = rootSegment.Subsegments.FirstOrDefault(s => s.Name == "SQL");
    if (sqlSegment != null) {
        recorder.SetEntity(sqlSegment);
    }
}
```

If I don't do this work around i get this exception:
```
AWSXRayRecorder.Instance.EndSubsegment()' threw an exception of type 'Amazon.XRay.Recorder.Core.Exceptions.EntityNotAvailableException
```
with inner exception:
```
System.InvalidCastException: Unable to cast object of type 'Amazon.XRay.Recorder.Core.Internal.Entities.Segment' to type 'Amazon.XRay.Recorder.Core.Internal.Entities.Subsegment'.
   at Amazon.XRay.Recorder.Core.AWSXRayRecorderImpl.PrepEndSubsegment()
   at Amazon.XRay.Recorder.Core.AWSXRayRecorderImpl.ProcessEndSubsegment()
   at Amazon.XRay.Recorder.Core.AWSXRayRecorder.EndSubsegment()
```

But this isn't how i expect the segments to work. I would expect the segments to behave like a stack, where BeginSubsegment() puts a segment on the stack, all annotations and metatdata get applied to that segment, until it is closed and it is popped making the parent the "current" entity.

All this is taking place within a web request using the xray middleware to setup the root segment.

Can anyone point out what I'm doing wrong, or as this seems like an odd edge as the docs suggest i'm doing it right... case perhaps i need to develop a PoC

**Comment by srprash** [comment]
Hi William,
Thanks for opening the issue. I would like to know some details about your code.
1. What is the xray middleware that you are using?
2. Are `ExecuteStart()` and `ExecuteFinish()` running in different threads?
 
If both of the above mentioned functions are in the same thread, then the `AWSXRayRecorder.Instance.GetEntity()` in `ExecuteFinish()` should get the `Subsegment` that is opened in the `ExecuteStart()`. 
On the other hand if the application is creating two separate threads for `ExecuteStart()` and `ExecuteFinish()`, then what I suspect is happening is that the Subsegment that you are creating in the `ExecuteStart()` thread is not available to the `ExecuteFinish()` thread and it has a stale TraceContext from the application this thread was spawned from.
Can you confirm this by trying to print the thread id for both the methods? 

Can you also try commenting out the async piece of code that you have mentioned in "Between Start and Finish Entity framework is executing an async db query, but no other profiling takes place." and see if that makes any difference?

Thanks,
Prashant

**Comment by williamdenton** [comment]
>What is the xray middleware that you are using?

The middleware supplied in this repo
https://github.com/aws/aws-xray-sdk-dotnet/blob/4ee0660ed6593b55a6c613f16201ffc58bc67b8f/sdk/src/Handlers/AspNetCore/AWSXRayMiddlewareExtensions.cs#L38-L41

---
>Are ExecuteStart() and ExecuteFinish() running in different threads?

Yes, they are on different threads. There must be a `ConfigureAwait(false)` somewhere in entity framework
This commit might not be the relevant place, but it does show that configureawait false is preferred in efcore codebase https://github.com/aspnet/EntityFrameworkCore/commit/5a48b4cfefac13c46c7b3f745b105ac27d9d106f

---

> commenting out the async piece of code

yes, converting the efcore DbSet call from async to sync results in the same thread being used and the entity being available as expected.

**Comment by srprash** [comment]
>Yes, they are on different threads. 

Ok. So how the .Net SDK for XRay in .NetCore works is that it maintains the `TraceContext` using [`AsyncLocal<Entity>`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.asynclocal-1?view=netframework-4.5) class object which is thread local. In your async code, when you create a subsegment in the `ExecuteStart()`, it is created in its own `TraceContext` and will not be available in the `ExecuteFinish()` thread's `TraceContext`. This will not be the case in a synchronous flow as you noticed when you changed the db call from sync to sync. 


> I'm using code like this (which i think is a gross work around):
```cs
var recorder = AWSXRayRecorder.Instance;
var segment = recorder.GetEntity();
if (segment is Amazon.XRay.Recorder.Core.Internal.Entities.Segment rootSegment) {
    var sqlSegment = rootSegment.Subsegments.FirstOrDefault(s => s.Name == "SQL");
    if (sqlSegment != null) {
        recorder.SetEntity(sqlSegment);
    }
}
```

But since the `rootSegment` is passed by reference to both the threads, you are able to retrieve the subsegment by traversing through the `rootSegment` the way you have mentioned above. 

---
>But this isn't how i expect the segments to work. I would expect the segments to behave like a stack, where BeginSubsegment() puts a segment on the stack, all annotations and metatdata get applied to that segment, until it is closed and it is popped making the parent the "current" entity. This works as a stack but only in a thread local memory space.

Although the segments/subsegments are not implemented as a stack, they do behave like a stack. When you begin a subsegment, it is linked to its parent entity and set in the TraceContext.
https://github.com/aws/aws-xray-sdk-dotnet/blob/bdb7471e2f14fc58da591321f08b3979729f0eb2/sdk/src/Core/AWSXRayRecorder.cs#L124-L128
And when you end a subsegment, the subsegment is pulled out of the TraceContext and its parent entity is then set in its place.
https://github.com/aws/aws-xray-sdk-dotnet/blob/bdb7471e2f14fc58da591321f08b3979729f0eb2/sdk/src/Core/AWSXRayRecorderImpl.cs#L855-L862

---

I think it might be a limitation from the .NetCore platform that operations on AsyncLocal from one thread is not reflected to other threads and your workaround is probably the way to go for now. 
I am keeping this issue as open and marking it for help. If anyone has faced a similar issue, they can suggest a way forward with this. 

Thanks,
Prashant

**Comment by williamdenton** [comment]
I believe this code will also be affected by this issue.

https://github.com/aws/aws-xray-sdk-dotnet/blob/bdb7471e2f14fc58da591321f08b3979729f0eb2/sdk/src/Core/AWSXRayRecorderImpl.cs#L939-L956

I am happy to make a PR to resolve this particular issue

**Comment by yogiraj07** [comment]
Hi @williamdenton ,
Sure PR is always welcomed. Can you help understand how `TraceMethodAsync()` code is affected?

**Comment by williamdenton** [comment]
I haven't had any luck writing a test that reproduces my problem. None of existing tests are specifically  testing resuming on a different thread. Modifying them to do so is trivial, but still no failures. I made the following changes...

Existing:
https://github.com/aws/aws-xray-sdk-dotnet/blob/f41ad03fbec63b63a246bf76e4748c5fe346cadf/sdk/test/UnitTests/AWSXRayRecorderTests.cs#L1113-L1117

Changed to:
```cs
        private async Task<int> PlusOneReturnAsync <TResult>(int count)
        {
            var threadId = System.Threading.Thread.CurrentThread.GetHashCode();
            await Task.Delay(1).ConfigureAwait(false);
            var resumedThreadId = System.Threading.Thread.CurrentThread.GetHashCode();
            Assert.AreNotEqual(threadId, resumedThreadId);
            return count +1;
        }
```

Existing:
https://github.com/aws/aws-xray-sdk-dotnet/blob/f41ad03fbec63b63a246bf76e4748c5fe346cadf/sdk/test/UnitTests/AWSXRayRecorderTests.cs#L1119-L1122

Changed to:
```cs
        private async Task PlusOneNoReturnAsync<TResult>( int count)
        {
            var threadId = System.Threading.Thread.CurrentThread.GetHashCode();
            await Task.Delay(1).ConfigureAwait(false);
            var resumedThreadId = System.Threading.Thread.CurrentThread.GetHashCode();
            Assert.AreNotEqual(threadId, resumedThreadId);
        }
```

for reference these little helpers are used in these tests
https://github.com/aws/aws-xray-sdk-dotnet/blob/f41ad03fbec63b63a246bf76e4748c5fe346cadf/sdk/test/UnitTests/AWSXRayRecorderTests.cs#L389-L416

Obviously these tests are not in the context of a aspnet app with entity framework etc but they seem to reproduce the problems that were suspected of being the source of the problem by @srprash 
Im not sure what to do from here...

**Comment by srprash** [comment]
HI @williamdenton ,
I appreciate you investigating the problem. Give me some time to understand what's happening here and I'll get back to you. Thanks!

**Comment by srprash** [comment]
Hi @williamdenton ,
I tested the changes you mentioned above in the context of an ASPNet, an ASPNetCore, and a .NET Console app. I did not run into any issues in all of those applications, and was able to generate the expected segments and subsegments.
I am a bit confused by your last comment in understanding that whether you were able to reproduce the problems through the above-mentioned code or did the tests pass for you?

We will keep looking into this issue on our end and would welcome any inputs.

Thanks.
